lib/gc/Transforms/Microkernel/MicrokernelInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,439 @@
+//===-- MicrokernelInvariantCodeMotion.cpp - Hoist invariance ---*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Rewrite/FrozenRewritePatternSet.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include <sstream>
+#include <utility>
+
+#include "gc/Transforms/Microkernel/BrgemmRuntimeUtils.h"
+#include "gc/Transforms/Microkernel/MicrokernelPasses.h"
+#include "gc/Transforms/Utils/ValueUtils.h"
+#include "oneapi/dnnl/dnnl_types.h"
+
+namespace mlir::microkernel {
+#define GEN_PASS_DEF_MICROKERNELINVARIANTCODEMOTION
+#include "gc/Transforms/Microkernel/MicrokernelPasses.h.inc"
+
+#define DEBUG_TYPE "microkernel-invariant-code-motion"
+
+enum BrgemmCallType { INAPPLICABLE = -1, DISPATCH, TILECFG, TILERELEASE };
+
+static bool isParallelLoop(Operation *op) {
+  return llvm::isa<scf::ForallOp>(op) || llvm::isa<scf::ParallelOp>(op) ||
+         llvm::isa<omp::ParallelOp>(op) || llvm::isa<omp::WsloopOp>(op);
+}
+
+static bool isConcernedCF(Operation *op) {
+  return llvm::isa<scf::ForOp>(op) || llvm::isa<scf::WhileOp>(op) ||
+         llvm::isa<scf::IfOp>(op) || llvm::isa<scf::IndexSwitchOp>(op);
+}
+
+static BrgemmCallType getBrgemmCallType(Operation *op) {
+  if (!llvm::isa<func::CallOp>(op)) {
+    return BrgemmCallType::INAPPLICABLE;
+  }
+  auto callOp = dyn_cast<func::CallOp>(op);
+  StringAttr callee = callOp.getCalleeAttr().getAttr();
+
+  if (callee == StringAttr::get(op->getContext(), DNNL_BRGEMM_DISPATCH_NAME)) {
+    return BrgemmCallType::DISPATCH;
+  }
+  if (callee == StringAttr::get(op->getContext(), DNNL_BRGEMM_TILECFG_NAME)) {
```

Comment:
can we extract the string in the callee instead of comparing it with a new stringattr?

Reply:
Fixed.